### ISSUE TITLE
Rewrite the whole thing

### ISSUE THREAD

**Issue by tobscure** [flarum/core#321]
_From @tobscure on August 27, 2015 13:51_

I threw this extension together in a night without much thought, because I wanted to get the beta out the door :P

While it does the basics OK (update the discussion list, push new posts, and update the unread notifications count), it could be way better. And it's a mess and basically needs to be rewritten. Anyway, here's how I think it should be improved:

* Extract a lot of the code into core, so that the extension is effectively telling core "hey I just found out about this new post in this discussion – whatchu gunna do about it?" and then core shows the "1 updated discussions" button or updates the post stream or whatever. This will allow us to write other extensions for different push services without duplicating/having inconsistent implementations.
* Make it push more information about the discussion that was updated, so that we can check whether we already have the most up-to-date information, and don't have to extend `addDiscussion`.
* Don't say there are updated discussions if they are in a hidden tag
* Push to a user whenever they read notifications, so that the number of unread notifications stays in sync between browser windows.
* Push for discussion events (e.g. discussion renaming)
* Push for post likes
* Show a little bubble every time a notification is received (again, this is something that would probably belong in core)
* Push user online status (via presence channels? idk)

What needs to be done:

* Working out a plan of attack for that first bullet point is the most important thing.
* Reimplement the current basic features in a cleaner way.
* Create issues for the additional features.

_Copied from original issue: flarum/pusher#1_

**Comment by franzliedke** [comment]
Also, the "new updates" notification in the discussion list [should be fixed to the top](http://discuss.flarum.org/d/629-show-updated-discussions-should-be-fixed).

**Comment by Luceos** [comment]
It would be amazing if the core would support a generic notification method and you can add new broadcasting systems to it. By default you would add pusher that would receive Events and handles them. But you could also think of e-mail, slack, hipchat; see for instance [notifymehq](https://github.com/notifymehq/notifyme).

**Comment by franzliedke** [comment]
Hmm, well, since we already have an event system, it's somewhat easy to do it for any notification provider anyway. But yeah, there might be value in abstracting away external services like that... could very well be done by an extension, though.

**Comment by Luceos** [comment]
Also take into consideration #512

**Comment by tarunmarkose** [comment]
Hey guys, quick question - will the rewritten Push system allow any extension to do a push event or only the core ones? 

For example, if I created a private messaging system, could it use your push extension to send out new messages as they came along? Kinda like a global push mechanism that any extension could tie into...

**Comment by Luceos** [comment]
@tarunmarkose any extension might use such an event system.

**Comment by Luceos** [comment]
@franzliedke @tobscure I'm considering picking this up, my proposal would be the following:

- Create a Message Interface and an implementation of that for flarum/core that will specify some basic properties: sender, receiver, title, body.
- Create a Broadcast Interface that specifies methods that are required to send Messages, eg: `send(MessageInterface $message)`
- Modify the Pusher package to use this methodology, implementing the Broadcast Interface.
- Allow extensions that have a Broadcast implementation to register themselves and make core send Messages over all registered Broadcasts. Any easy way of doing this? Maybe a seperate event, RegisterBroadcaster?

Also I agree the pusher rewrite is a seperate issue related to this, but requires this to be completed.

In hindsight, Channel (or MessageChannel) over Broadcast.

Awaiting feedback.

**Comment by franzliedke** [comment]
I'd say we need separate issues for most of these.

First step: Implementing the interfaces.
Second step: A manager for broadcasts (possibly one implementation of the interface that has references to all other instances). I can take care of that.
Third step: Adapting the extension.

**Comment by tobscure** [comment]
Sounds good to me.

**Comment by Luceos** [comment]
This might be great solution for the manager: https://github.com/notifymehq/notifyme @franzliedke

**Comment by tobscure** [comment]
Hmm, I just had a bit of a brainstorm about how to better factor in discussion/tag permissions when pushing new posts. The currently implementation is very basic: if the post/discussion is visible to guests, push it to the public channel, otherwise don't push it at all. That's not very useful for members-only forums though!

It's a complex problem though, because Flarum's permissions are so dynamic. We could add code in the push mechanism to check which tags a discussion has and which groups are allowed to view them... but then what if a forum isn't using the tags extensions, and is using something else instead? And how would a private discussion extension fit in? Hardcoding references to extensions and their logic is bad.

A potential solution to this would be to add a couple of extensible methods to the Post model which return which groups/users that the post is visible to. Kind of like a reverse permission lookup – instead of "can this user (in these groups) view this post?" it's asking "this post, which users/groups can view it?"

By default it would just be implemented something like this:

```php
public function groupsThatCanView() {
    return Permission::where('permission', 'viewDiscussions')->lists('group_id')->all();
}
public function usersThatCanView() {
    return []; // not visible to any users in particular
}
```

But then there would also be a couple of new events in there for extensions to hook into. Tags would do something like this:

```php
$events->listen(PrepareGroupsThatCanViewPost::class, function ($event) {
    // Loop through each of the discussion's tags. If any of them are restricted,
    // then determine which groups can view that tag and return that.
    foreach ($event->post->discussion->tags as $tag) {
        if ($tag->is_restricted) {
            $event->groups = Permission::where('permission', 'tag'.$tag->id.'.viewDiscussions')->lists('group_id')->all();
        }
    }
});
```

And the private discussions extension would do:

```php
$events->listen(PrepareUsersThatCanViewPost::class, function ($event) {
    // Get the IDs of all the users that are invited to this private discussion
    $event->users = $event->post->discussion->privateUsers->lists('id')->all();
});
```

**Comment by franzliedke** [comment]
What about pushing an event on any activity, and then polling for new posts from the client once those events are received? That's obviously a very naive solution...

**Comment by Luceos** [comment]
Polling would reduce the benefit of the websocket implementation al together @franzliedke . Even so, your remark does make sense and is way less complicated than what @tobscure proposes. The question is, how to push all events and filter them properly by rights before broadcasting.

The solution by @tobscure is practical, even though it's a bit complicated to understand at first. It might actually work well. I do see only examples of giving access, how would one remove access through an event listener?

**Comment by franzliedke** [comment]
Yeah, bad wording again. It's not really "polling", just loading new data exactly once when receiving an event. That would probably cause a lot of unnecessary traffic on the server...

**Comment by tobscure** [comment]
> @franzliedke What about pushing an event on any activity, and then polling for new posts from the client once those events are received? That's obviously a very naive solution...

That would be a nice solution, though it's not ideal security-wise... everyone would know exactly when and how many posts are being made across the whole forum, even in sections they don't know exist... not a huge deal but still worth mentioning. And yeah, that could mean a lot of unnecessary traffic.

> @Luceos  I do see only examples of giving access, how would one remove access through an event listener?

The PrepareGroupsThatCanViewPost event would have an array of groups passed by reference, so items in `$event->groups` can be modified/added/removed as needed.

**Comment by Luceos** [comment]
I'm starting to feel a lot for the reverse permission logic @tobscure . It might be useful in other areas as well in the near future. Also I don't think it's wise to broadcast all events to everyone, who knows what information extensions will be sending along at some point.

**Comment by Luceos** [comment]
I'm removing myself from this task until I have the chance to actually start the implementation.

**Comment by tobscure** [comment]
> Also I don't think it's wise to broadcast all events to everyone, who knows what information extensions will be sending along at some point.

Right, that's the biggest concern and I think it justifies the implementation of reverse permission logic.

**Comment by Luceos** [comment]
I was under the impression we already decided on that @tobscure :laughing:

**Comment by frostymarvelous** [comment]
There is a bug here.

    $event->groups = Permission::where('permission', 'tag'.$tag->id.'.viewDiscussions')->lists('group_id')->all();
       
What happens when the next tag in the loop is restricted as well? Isn't `$event->groups` overwritten?
Sorry, I'm treating my glaucoma so I can't trust my reasoning.

**Comment by Luceos** [comment]
@frostymarvelous yes so the implementation should have an array like `$requiredGroup` which is filled throughout the loop. Not that much of an issue.

**Comment by mauricederegt** [comment]
To be honest, why not keep it as a plugin? Pusher is expensive and a paid service most users will not use. Keeping this as a plugin will keep Flarum small and clean and even better: keeps the option open to use different (or own setup) pusher like services via a plugin (and this forum is all about plugins right?)

**Comment by frostymarvelous** [comment]
From the original post.

'Extract a lot of the code into core, so that the extension is effectively telling core "hey I just found out about this new post in this discussion – whatchu gunna do about it?" and then core shows the "1 updated discussions" button or updates the post stream or whatever. This will allow us to write other extensions for different push services without duplicating/having inconsistent implementations.'

The aspects being included in the core is essentially a notification event. And then we can write different plugins to interact with this event to enhance the experience.

**Comment by Luceos** [comment]
Note that the broadcasting implementation needs to be separated from this issue into three other issues:

- interface design for broadcasting
- implementation of a broadcasting manager
- example by modifying the mentions extension and possibly pusher extension

**Comment by Luceos** [comment]
Laravel 5.2 offers the illuminate broadcasting package. Might be interesting for this implementation:

https://github.com/illuminate/broadcasting
